scipy/stats/_qmc.py

```diff
@@ -243,80 +254,44 @@ def discrepancy(sample, iterative=False, method='CD'):
-            prod_arr *= (1
-                         + 0.5 * abs(s0[:, None] - 0.5) + 0.5 * abs(s0 - 0.5)
-                         - 0.5 * abs(s0[:, None] - s0))
-        disc2 = prod_arr.sum()
-
-        return ((13.0 / 12.0) ** d - 2.0 / n * disc1 +
-                1.0 / (n ** 2) * disc2)
-    elif method == 'WD':
-        # reference [1], page 73 Eq (3.8)
-        prod_arr = 1
-        for i in range(d):
-            s0 = sample[:, i]
-            x_kikj = abs(s0[:, None] - s0)
-            prod_arr *= 3.0 / 2.0 - x_kikj + x_kikj ** 2
-
-        # typo in the book sign missing: - (4.0 / 3.0) ** d
-        return - (4.0 / 3.0) ** d + 1.0 / (n ** 2) * prod_arr.sum()
-    elif method == 'MD':
-        # reference [2], page 290 Eq (18)
-        abs_ = abs(sample - 0.5)
-        disc1 = np.sum(np.prod(5.0 / 3.0 - 0.25 * abs_ - 0.25 * abs_ ** 2,
-                               axis=1))
-
-        prod_arr = 1
-        for i in range(d):
-            s0 = sample[:, i]
-            prod_arr *= (15.0 / 8.0
-                         - 0.25 * abs(s0[:, None] - 0.5) - 0.25 * abs(s0 - 0.5)
-                         - 3.0 / 4.0 * abs(s0[:, None] - s0)
-                         + 0.5 * abs(s0[:, None] - s0) ** 2)
-        disc2 = prod_arr.sum()
-
-        disc = (19.0 / 12.0) ** d
-        disc1 = 2.0 / n * disc1
-        disc2 = 1.0 / (n ** 2) * disc2
-
-        return disc - disc1 + disc2
-    elif method == 'L2-star':
-        # reference [1], page 69 Eq (3.5)
-        disc1 = np.sum(np.prod(1 - sample**2, axis=1))
-
-        xik = sample[None, :, :]
-        xjk = sample[:, None, :]
-        disc2 = np.sum(np.sum(np.prod(1 - np.maximum(xik, xjk), axis=2),
-                              axis=1))
-
-        return np.sqrt(
-            3 ** (-d) - 1 / n * 2 ** (1 - d) * disc1 + 1 / (n ** 2) * disc2
+    if not isinstance(workers, int):
+        raise ValueError(
+            f"workers param needs to be an int. Found " f"{type(workers)!r}"
         )
+
+    if workers == -1:
+        workers = os.cpu_count()
+        if workers is None:
+            raise NotImplementedError(
+                "Cannot determine the number of cpus using os.cpu_count(), "
+                "cannot use -1 for the number of workers")
+    elif workers <= 0:
+        raise ValueError(f"Invalid number of workers: {workers}, must be -1 "
+                         "or > 0")
+
+    methods = {
+        "CD": _cy_wrapper_centered_discrepancy,
+        "WD": _cy_wrapper_wrap_around_discrepancy,
+        "MD": _cy_wrapper_mixture_discrepancy,
+        "L2-star": _cy_wrapper_l2_star_discrepancy,
+    }
+
+    if method in methods:
+        return methods[method](sample, iterative, workers=workers)
     else:
-        raise ValueError('{} is not a valid method. Options are '
-                         'CD, WD, MD, L2-star.'.format(method))
+        raise ValueError(
+            "{!r} is not a valid method. Options are "
+            "CD, WD, MD, L2-star.".format(method)
+        )
```

Comment:
Now you have a `methods` dict. The name of the methods should be deduced from keys.

```suggestion
        raise ValueError(
            f"{method!r} is not a valid method. Options are "
            + ", ".join(f"{x!r}" for x in methods.keys()) "."
        )
```

Reply:
I might be misunderstanding the code, but perhaps `f"{method} is not a valid method. It must be one of {set(methods)}"` is slightly more compact?

Reply:
I agree, it adds braces. but it is not a problem (and I prefer with braces).

However, I prefer keep the `!r` in `{method!r}`.

Reply:
Thank you for your feedbacks. I updated it in c099d5400764d1bd184532e19ca7effc3afa4d67. I wonder if the changes in tests suit all of you ?

Reply:
Perfect for me.